lightningd/pay.c

```diff
@@ -1013,74 +1042,12 @@ static struct command_result *json_sendonion(struct command *cmd,
 				    "with failcode=%d",
 				    failcode);
 
-	/* Now, do we already have a payment? */
-	payment = wallet_payment_by_hash(tmpctx, ld->wallet, payment_hash, /* FIXME: Set parallel_id! */0);
-	if (payment) {
-		if (payment->status == PAYMENT_PENDING) {
-			log_debug(ld->log, "send_payment: previous still in progress");
-			return json_sendpay_in_progress(cmd, payment);
-		}
-		if (payment->status == PAYMENT_COMPLETE) {
-			log_debug(ld->log, "send_payment: previous succeeded");
-			return sendpay_success(cmd, payment);
-		}
-		log_debug(ld->log, "send_payment: found previous, retrying");
-	}
-
-	channel = active_channel_by_id(ld, &first_hop->nodeid, NULL);
-	if (!channel) {
-		struct json_stream *data
-			= json_stream_fail(cmd, PAY_TRY_OTHER_ROUTE,
-					   "No connection to first "
-					   "peer found");
-
-		json_add_routefail_info(data, 0, WIRE_UNKNOWN_NEXT_PEER,
-					&ld->id, &first_hop->channel_id,
-					node_id_idx(&ld->id, &first_hop->nodeid),
-					NULL);
-		json_object_end(data);
-		return command_failed(cmd, data);
-	}
-
-	/* Cleanup any prior payment. We're about to retry. */
-	if (payment) {
-		wallet_payment_delete(ld->wallet, payment_hash, /* FIXME: Set parallel_id! */0);
-		wallet_local_htlc_out_delete(ld->wallet, channel, payment_hash, /* FIXME: Set parallel_id! */0);
-	}
-
-	failcode = send_onion(cmd->ld, packet, first_hop, payment_hash, /* FIXME: Set parallel_id! */0, channel,
-			  &hout);
-
-	payment = tal(hout, struct wallet_payment);
-	payment->id = 0;
-	payment->payment_hash = *payment_hash;
-	payment->status = PAYMENT_PENDING;
-	payment->msatoshi = AMOUNT_MSAT(0);
-	payment->msatoshi_sent = first_hop->amount;
-	payment->timestamp = time_now().ts.tv_sec;
-
-	/* These are not available for sendonion payments since the onion is
-	 * opaque and we can't extract them. Errors have to be handled
-	 * externally, since we can't decrypt them.*/
-	payment->destination = NULL;
-	payment->payment_preimage = NULL;
-	payment->route_nodes = NULL;
-	payment->route_channels = NULL;
-	payment->bolt11 = NULL;
-	payment->failonion = NULL;
-	payment->path_secrets = tal_steal(payment, path_secrets);
-
-	if (label != NULL)
-		payment->label = tal_strdup(payment, label);
-	else
-		payment->label = NULL;
-
-	/* We write this into db when HTLC is actually sent. */
-	wallet_payment_setup(ld->wallet, payment);
-
-	add_sendpay_waiter(ld, cmd, payment_hash);
-	return command_still_pending(cmd);
+	return send_payment_core(ld, cmd, payment_hash, /* FIXME: Set parallel_id! */0,
+				 first_hop, AMOUNT_MSAT(0), AMOUNT_MSAT(0),
```

Comment:
Pretty sure setting `msat_total = AMOUNT_MSAT(0)` on all onion payments will cause issues further down the line. We might need to have an exemption in the check in the next commit to allow payments with different onions to exceed the `msatoshi_total` amount.